File: benchmarks/compute/nbody/Go/nbody.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"time"
)

// ---------- 可选数据规模 ----------
const (
	SMALL     = "SMALL"
	STANDARD  = "STANDARD"
	LARGE     = "LARGE"
	DATASET   = STANDARD
)

var (
	BODIES      int
	ITERATIONS  int
)

func init() {
	switch DATASET {
	case SMALL:
		BODIES = 100
		ITERATIONS = 10
	case LARGE:
		BODIES = 5000
		ITERATIONS = 100
	default:
		BODIES = 1000
		ITERATIONS = 50
	}
}

type Vec3 struct {
	x, y, z float64
}

type Body struct {
	position    Vec3
	velocity    Vec3
	acceleration Vec3
	mass        float64
}

// ---------- 初始化天体 ----------
func initBodies(n int, seed int64) []Body {
	rand.Seed(seed)
	bodies := make([]Body, n)
	for i := 0; i < n; i++ {
		bodies[i] = Body{
			position: Vec3{
				x: rand.Float64() * 100.0,
				y: rand.Float64() * 100.0,
				z: rand.Float64() * 100.0,
			},
			velocity: Vec3{
				x: rand.Float64() * 10.0,
				y: rand.Float64() * 10.0,
				z: rand.Float64() * 10.0,
			},
			acceleration: Vec3{
				x: 0.0,
				y: 0.0,
				z: 0.0,
			},
			mass: rand.Float64()*1000.0 + 100.0,
		}
	}
	return bodies
}

// ---------- 打印部分输出 ----------
func printResult(bodies []Body) {
	for i := 0; i < 3 && i < len(bodies); i++ {
		fmt.Printf("Body %d: pos=(%.2f, %.2f, %.2f) vel=(%.2f, %.2f, %.2f)\n",
			i, bodies[i].position.x, bodies[i].position.y, bodies[i].position.z,
			bodies[i].velocity.x, bodies[i].velocity.y, bodies[i].velocity.z)
	}
}

// ---------- 主计算过程 ----------
func computeNbody(n, iterations int, bodies []Body) {
	const G = 6.67430e-11 // 万有引力常数

	for iter := 0; iter < iterations; iter++ {
		// 重置加速度
		for i := 0; i < n; i++ {
			bodies[i].acceleration.x = 0.0
			bodies[i].acceleration.y = 0.0
			bodies[i].acceleration.z = 0.0
		}

		// 计算引力
		for i := 0; i < n; i++ {
			for j := i + 1; j < n; j++ {
				dx := bodies[j].position.x - bodies[i].position.x
				dy := bodies[j].position.y - bodies[i].position.y
				dz := bodies[j].position.z - bodies[i].position.z

				distSq := dx*dx + dy*dy + dz*dz + 1e-10 // 避免除以零
				dist := math.Sqrt(distSq)
				force := G * bodies[i].mass * bodies[j].mass / distSq

				fx := force * dx / dist
				fy := force * dy / dist
				fz := force * dz / dist

				bodies[i].acceleration.x += fx / bodies[i].mass
				bodies[i].acceleration.y += fy / bodies[i].mass
				bodies[i].acceleration.z += fz / bodies[i].mass

				bodies[j].acceleration.x -= fx / bodies[j].mass
				bodies[j].acceleration.y -= fy / bodies[j].mass
				bodies[j].acceleration.z -= fz / bodies[j].mass
			}
		}

		// 更新速度和位置
		for i := 0; i < n; i++ {
			bodies[i].velocity.x += bodies[i].acceleration.x
			bodies[i].velocity.y += bodies[i].acceleration.y
			bodies[i].velocity.z += bodies[i].acceleration.z

			bodies[i].position.x += bodies[i].velocity.x
			bodies[i].position.y += bodies[i].velocity.y
			bodies[i].position.z += bodies[i].velocity.z
		}
	}
}

// ---------- 主函数 ----------
func main() {
	var seed int64 = 42
	if len(os.Args) > 1 {
		if s, err := strconv.ParseInt(os.Args[1], 10, 64); err == nil {
			seed = s
		}
	}

	bodies := initBodies(BODIES, seed)

	start := time.Now()
	computeNbody(BODIES, ITERATIONS, bodies)
	duration := time.Since(start)

	fmt.Printf("Finished N-body simulation in %.3f seconds\n", duration.Seconds())
	printResult(bodies)
}

File: benchmarks/compute/nbody/C/nbody.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define SMALL 0
#define STANDARD 1
#define LARGE 2

// ---------- 可选数据规模 ----------
#if !defined(DATASET)
#define DATASET STANDARD
#endif

#if DATASET == SMALL
#define BODIES 100
#define ITERATIONS 10
#elif DATASET == LARGE
#define BODIES 5000
#define ITERATIONS 100
#else
#define BODIES 1000
#define ITERATIONS 50
#endif

typedef struct {
    double x, y, z;
} Vec3;

typedef struct {
    Vec3 position;
    Vec3 velocity;
    Vec3 acceleration;
    double mass;
} Body;

// ---------- 初始化天体 ----------
void init_bodies(int n, Body bodies[n], unsigned int seed) {
    srand(seed);
    for (int i = 0; i < n; i++) {
        bodies[i].position.x = (double)rand() / RAND_MAX * 100.0;
        bodies[i].position.y = (double)rand() / RAND_MAX * 100.0;
        bodies[i].position.z = (double)rand() / RAND_MAX * 100.0;
        
        bodies[i].velocity.x = (double)rand() / RAND_MAX * 10.0;
        bodies[i].velocity.y = (double)rand() / RAND_MAX * 10.0;
        bodies[i].velocity.z = (double)rand() / RAND_MAX * 10.0;
        
        bodies[i].acceleration.x = 0.0;
        bodies[i].acceleration.y = 0.0;
        bodies[i].acceleration.z = 0.0;
        
        bodies[i].mass = (double)rand() / RAND_MAX * 1000.0 + 100.0;
    }
}

// ---------- 打印部分输出（防止被编译器优化） ----------
void print_result(int n, Body bodies[n]) {
    for (int i = 0; i < n && i < 3; i++) {
        printf("Body %d: pos=(%.2f, %.2f, %.2f) vel=(%.2f, %.2f, %.2f)\n",
               i, bodies[i].position.x, bodies[i].position.y, bodies[i].position.z,
               bodies[i].velocity.x, bodies[i].velocity.y, bodies[i].velocity.z);
    }
}

// ---------- 主计算过程 ----------
void compute_nbody(int n, int iterations, Body bodies[n]) {
    const double G = 6.67430e-11; // 万有引力常数
    
    for (int iter = 0; iter < iterations; iter++) {
        // 重置加速度
        for (int i = 0; i < n; i++) {
            bodies[i].acceleration.x = 0.0;
            bodies[i].acceleration.y = 0.0;
            bodies[i].acceleration.z = 0.0;
        }
        
        // 计算引力
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                double dx = bodies[j].position.x - bodies[i].position.x;
                double dy = bodies[j].position.y - bodies[i].position.y;
                double dz = bodies[j].position.z - bodies[i].position.z;
                
                double dist_sq = dx*dx + dy*dy + dz*dz + 1e-10; // 避免除以零
                double dist = sqrt(dist_sq);
                double force = G * bodies[i].mass * bodies[j].mass / dist_sq;
                
                double fx = force * dx / dist;
                double fy = force * dy / dist;
                double fz = force * dz / dist;
                
                bodies[i].acceleration.x += fx / bodies[i].mass;
                bodies[i].acceleration.y += fy / bodies[i].mass;
                bodies[i].acceleration.z += fz / bodies[i].mass;
                
                bodies[j].acceleration.x -= fx / bodies[j].mass;
                bodies[j].acceleration.y -= fy / bodies[j].mass;
                bodies[j].acceleration.z -= fz / bodies[j].mass;
            }
        }
        
        // 更新速度和位置
        for (int i = 0; i < n; i++) {
            bodies[i].velocity.x += bodies[i].acceleration.x;
            bodies[i].velocity.y += bodies[i].acceleration.y;
            bodies[i].velocity.z += bodies[i].acceleration.z;
            
            bodies[i].position.x += bodies[i].velocity.x;
            bodies[i].position.y += bodies[i].velocity.y;
            bodies[i].position.z += bodies[i].velocity.z;
        }
    }
}

// ---------- 主函数 ----------
int main(int argc, char** argv) {
    static Body bodies[BODIES];
    unsigned int seed = (argc > 1) ? atoi(argv[1]) : 42;

    init_bodies(BODIES, bodies, seed);

    clock_t start = clock();
    compute_nbody(BODIES, ITERATIONS, bodies);
    clock_t end = clock();

    printf("Finished N-body simulation in %.3f seconds\n",
           (double)(end - start) / CLOCKS_PER_SEC);

    print_result(BODIES, bodies);
    return 0;
}
